core/src/main/scala/com/salesforce/op/utils/stages/NameDetectUtils.scala

```diff
@@ -0,0 +1,377 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.salesforce.op.utils.stages
+
+import com.salesforce.op.features.types.{NameStats, Text}
+import com.salesforce.op.stages.impl.feature.TextTokenizer
+import com.salesforce.op.utils.json.{JsonLike, JsonUtils}
+import com.twitter.algebird._
+import com.twitter.algebird.Operators._
+import com.twitter.algebird.macros.caseclass._
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.internal.Logging
+import org.apache.spark.ml.param.{BooleanParam, DoubleParam, ParamValidators, Params}
+
+import scala.io.Source
+import scala.util.Try
+import scala.util.matching.Regex
+
+/**
+ * Provides shared helper functions and variables for name identification and name to gender transformation.
+ * @tparam T     the FeatureType (subtype of Text) to operate over
+ */
+private[op] trait NameDetectFun[T <: Text] extends NameDetectParams with Logging {
+  import NameDetectUtils._
+
+  def preProcess(input: T#Value): Seq[String] = {
+    TextTokenizer.tokenize(Text(input)).tokens.toArray
+  }
+
+  def computeGuardCheckQuantities(
+    text: String,
+    tokens: Seq[String],
+    hllMonoid: HyperLogLogMonoid
+  ): GuardCheckStats = {
+    val textLength = text.length
+    GuardCheckStats(
+      countBelowMaxNumTokens = if (tokens.length < 10) 1 else 0,
+      countAboveMinCharLength = if (textLength > 2) 1 else 0,
+      approxMomentsOfTextLength = Moments(textLength),
+      approxNumUnique = hllMonoid.create(text.getBytes)
+    )
+  }
+
+  def performGuardChecks(stats: GuardCheckStats, hllMonoid: HyperLogLogMonoid): Boolean = {
+    val N: Double = stats.approxMomentsOfTextLength.count.toDouble
+    val checks = List(
+      // check that in at least 3/4 of the texts there are no more than 10 tokens
+      (stats.countBelowMaxNumTokens / N) > 0.75,
+      // check that at least 3/4 of the texts are longer than 3 characters
+      (stats.countAboveMinCharLength / N) > 0.75,
+      // check that the standard deviation of the text length is greater than a small number
+      N < 10 || stats.approxMomentsOfTextLength.stddev > 0.05,
+      // check that the number of unique entries is at least 10
+      N < 100 || hllMonoid.sizeOf(stats.approxNumUnique).estimate > 10
+    )
+    checks.forall(identity)
+  }
+
+  def dictCheck(tokens: Seq[String], dict: Broadcast[NameDictionary]): Double = {
+    if (tokens.length == 0) 0.0 else {
+      tokens.map({ token: String => if (dict.value.value contains token) 1 else 0}).sum.toDouble / tokens.length
+    }
+  }
+
+  def getNameFromCustomIndex(tokens: Seq[String], index: Int): String = {
+    if (tokens.isEmpty) ""
+    else if (tokens.length == 1) tokens.head
+    else {
+      // Mod to accept -1 as valid index
+      tokens((index + tokens.length) % tokens.length)
```

Comment:
This might just be being too Pythonic in Scala-land but I wanted functionality for getting the last element in a list. I could make this more explicit in the `GenderDetectStrategy` enum by making `GetFirst` and `GetLast` options but I thought it would be more useful to just specify it in a `ByIndex` class so that somebody could specify arbitrary indices to check.